scripting/lua/api/BonusSystem.cpp

```diff
@@ -0,0 +1,255 @@
+{
+	LuaStack S(L);
+	std::shared_ptr<const Bonus> object;
+	if(!S.tryGet(1, object))
+		return S.retNil();
+	return LuaStack::quickRetInt(L, object->effectRange);
+}
+
+int BonusProxy::getStacking(lua_State * L)
+{
+	LuaStack S(L);
+	std::shared_ptr<const Bonus> object;
+	if(!S.tryGet(1, object))
+		return S.retNil();
+	return LuaStack::quickRetStr(L, object->stacking);
+}
+
+int BonusProxy::getDescription(lua_State * L)
+{
+	LuaStack S(L);
+	std::shared_ptr<const Bonus> object;
+	if(!S.tryGet(1, object))
+		return S.retNil();
+	return LuaStack::quickRetStr(L, object->description);
+}
+
+int BonusProxy::toJsonNode(lua_State * L)
+{
+	LuaStack S(L);
+	std::shared_ptr<const Bonus> object;
+	if(!S.tryGet(1, object))
+		return S.retNil();
+	S.clear();
+	S.push(object->toJsonNode());
+	return 1;
+}
+
+template <typename T>
+static void publishMap(lua_State * L, const T & map)
+{
+	for(auto & p : map)
+	{
+		const std::string & name = p.first;
+		int32_t id = static_cast<int32_t>(p.second);
+
+		lua_pushstring(L, name.c_str());
+		lua_pushinteger(L, id);
+		lua_rawset(L, -3);
+	}
+}
+
+void BonusProxy::adjustStaticTable(lua_State * L) const
+{
+	publishMap(L, bonusNameMap);
+	publishMap(L, bonusValueMap);
+	publishMap(L, bonusSourceMap);
+	publishMap(L, bonusDurationMap);
+}
+
+VCMI_REGISTER_SCRIPT_API(BonusListProxy, "BonusList");
+
+const std::vector<BonusListProxy::CustomRegType> BonusListProxy::REGISTER_CUSTOM =
+{
+
+};
+
+std::shared_ptr<const Bonus> BonusListProxy::index(std::shared_ptr<const BonusList> self, int key)
+{
+	//field = __index(self, key)
+
+	std::shared_ptr<const Bonus> ret;
+
+	if((key >= 1) && (key <= self->size()))
+		ret = (*self)[key-1];
+	return ret;
+}
+
+void BonusListProxy::adjustMetatable(lua_State * L) const
+{
+	lua_pushstring(L, "__index");
+	lua_pushcclosure(L, LuaFunctionWrapper<decltype(&BonusListProxy::index), &BonusListProxy::index>::invoke, 0);
+	lua_rawset(L, -3);
+}
+
+VCMI_REGISTER_SCRIPT_API(BonusBearerProxy, "BonusBearer");
+
+const std::vector<BonusBearerProxy::CustomRegType> BonusBearerProxy::REGISTER_CUSTOM =
+{
+	{"getBonuses", &BonusBearerProxy::getBonuses, false},
+};
+
+int BonusBearerProxy::getBonuses(lua_State * L)
+{
+	LuaStack S(L);
+
+	const IBonusBearer * object = nullptr;
+	if(!S.tryGet(1, object))
+		return S.retNil();
+
+	TConstBonusListPtr ret;
+
+	const bool hasSelector = S.isFunction(2);
+	const bool hasRangeSelector = S.isFunction(3);
+
+	if(hasSelector)
+	{
+		auto selector = [](const Bonus * b)
+		{
+			return false; //TODO: BonusBearerProxy::getBonuses selector
+		};
+
+		if(hasRangeSelector)
+		{
+			auto rangeSelector = [](const Bonus * b)
+			{
+				return false;//TODO: BonusBearerProxy::getBonuses rangeSelector
+			};
+
+			ret = object->getBonuses(selector, rangeSelector);
+		}
+		else
+		{
+			ret = object->getBonuses(selector, Selector::all);
+		}
+	}
+	else
+	{
+		ret = object->getBonuses(Selector::all, Selector::all);
+	}
+
+	S.clear();
+	S.push(ret);
+	return S.retPushed();
```

Comment:
In the other cases I have seen return 1; Is it some special case?

Reply:
In this case `return 1` is also correct, but retPushed also works for "multy-return". The return value is always a number of returned values on the stack.